core/src/main/java/jenkins/tasks/SimpleBuildStep.java

```diff
@@ -80,9 +82,37 @@
      * @param listener a place to send output
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @deprecated Use {@link #perform(Run, FilePath, EnvVars, Launcher, TaskListener) instead.}
      */
-    void perform(@NonNull Run<?,?> run, @NonNull FilePath workspace, @NonNull Launcher launcher,
-                 @NonNull TaskListener listener) throws InterruptedException, IOException;
+    @Deprecated
+    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull Launcher launcher,
+                         @NonNull TaskListener listener) throws InterruptedException, IOException {
+        // No additional environment available; just use that from the Run.
+        this.perform(run, workspace, run.getEnvironment(listener), launcher, listener);
+    }
+
+    /**
+     * Run this step.
+     * @param run a build this is running as a part of
+     * @param workspace a workspace to use for any file operations
+     * @param env environment variables applicable to this step
+     * @param launcher a way to start processes
+     * @param listener a place to send output
+     * @throws InterruptedException if the step is interrupted
+     * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @since 2.2xx
+     */
+    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull EnvVars env, @NonNull Launcher launcher,
+                         @NonNull TaskListener listener) throws InterruptedException, IOException {
+        // If this is called, this must be an implementer of the previous API, in which case we call that, discarding
+        // the environment we were given.
+        // But for that to work, that API method must have been implemented.
+        if (!Util.isOverridden(SimpleBuildStep.class, this.getClass(),
+                "perform", Run.class, FilePath.class, Launcher.class, TaskListener.class)) {
+            throw new AbstractMethodError();
+        }
+        this.perform(run, workspace, launcher, listener);
```

Comment:
```suggestion
        if (Util.isOverridden(SimpleBuildStep.class, this.getClass(),
                "perform", Run.class, FilePath.class, Launcher.class, TaskListener.class)) {
            this.perform(run, workspace, launcher, listener);
        } else {
            throw new AbstractMethodError();
        }
```

may be easier to follow